src/hipoc/hipoc_kernel.cpp

```diff
@@ -123,9 +125,81 @@ void HIPOCKernelInvoke::run(void* args, std::size_t size) const
     }
 }
 
+void HIPOCKernelInvoke::run_cooperative(void** kern_args) const
+{
+    hipError_t status;
+
+    MIOPEN_LOG_I2("kernel_name = "
+                  << GetName() << ", global_work_dim = " << DimToFormattedString(gdims.data(), 3)
+                  << ", local_work_dim = " << DimToFormattedString(ldims.data(), 3));
+
+    const auto& arch = miopen::GetStringEnv(ENV(MIOPEN_DEVICE_ARCH));
+    if(!arch.empty())
+    {
+        MIOPEN_THROW("MIOPEN_DEVICE_ARCH used, escaping launching kernel");
+    }
+
+    HipEventPtr start = nullptr;
+    HipEventPtr stop  = nullptr;
+
+    if(callback)
+    {
+        start = make_hip_event();
+        stop  = make_hip_event();
+    }
+
+#if WORKAROUND_SWDEV_448157
+    if(gdims[0] >= (1ULL << 32) || gdims[1] >= (1ULL << 32) || gdims[2] >= (1ULL << 32))
+        MIOPEN_THROW("gridDim x blockDim >= 2^32");
+
+    if(gdims[0] % ldims[0] != 0 || gdims[1] % ldims[1] != 0 || gdims[2] % ldims[2] != 0)
+        MIOPEN_THROW(miopenStatusInternalError);
+
+    unsigned grid_dim_x = gdims[0] / ldims[0];
+    unsigned grid_dim_y = gdims[1] / ldims[1];
+    unsigned grid_dim_z = gdims[2] / ldims[2];
+
+    MIOPEN_HANDLE_LOCK
+
+    if(callback)
```

Comment:
this should probably go before #if

Reply:
@DrizztDoUrden This is part of the workaround. `hipModuleLaunchCooperativeKernel()` does not work with events, so I had to do it this way.